openhands/integrations/bitbucket/bitbucket_service.py

```diff
@@ -0,0 +1,371 @@
+    """
+
+    BASE_URL = 'https://api.bitbucket.org/2.0'
+    token: SecretStr = SecretStr('')
+    refresh = False
+
+    def __init__(
+        self,
+        user_id: str | None = None,
+        external_auth_id: str | None = None,
+        external_auth_token: SecretStr | None = None,
+        token: SecretStr | None = None,
+        external_token_manager: bool = False,
+        base_domain: str | None = None,
+    ):
+        self.user_id = user_id
+        self.external_token_manager = external_token_manager
+        self.external_auth_id = external_auth_id
+        self.external_auth_token = external_auth_token
+        self.base_domain = base_domain or 'bitbucket.org'
+
+        if token:
+            self.token = token
+        elif os.environ.get('BITBUCKET_TOKEN'):
+            self.token = SecretStr(os.environ.get('BITBUCKET_TOKEN', ''))
+
+    @property
+    def provider(self) -> str:
+        return 'Bitbucket'
+
+    async def get_latest_token(self) -> SecretStr | None:
+        """Get latest working token of the user."""
+        if self.external_token_manager and self.external_auth_id:
+            # This would be implemented by a custom token manager
+            return None
+        return self.token
+
+    async def _make_request(
+        self,
+        url: str,
+        params: dict | None = None,
+        method: RequestMethod = RequestMethod.GET,
+    ) -> tuple[Any, dict]:
+        """Make a request to the Bitbucket API.
+
+        Args:
+            url: The URL to request
+            params: Optional parameters for the request
+            method: The HTTP method to use
+
+        Returns:
+            A tuple of (response_data, response_headers)
+
+        Raises:
+            AuthenticationError: If the token is invalid
+            RateLimitError: If the rate limit is exceeded
+            UnknownException: For other errors
+        """
+        # Bitbucket Cloud API supports both Bearer token and Basic auth
+        # For app passwords, we need to use Basic auth with the format username:app_password
+        token_value = self.token.get_secret_value()
+
+        # Check if the token contains a colon, which indicates it's in username:password format
+        if ':' in token_value:
+            import base64
+
+            auth_str = base64.b64encode(token_value.encode()).decode()
+            headers = {
+                'Authorization': f'Basic {auth_str}',
+                'Accept': 'application/json',
+            }
+        else:
+            headers = {
+                'Authorization': f'Bearer {token_value}',
+                'Accept': 'application/json',
+            }
+
+        try:
+            async with httpx.AsyncClient() as client:
+                response = await self.execute_request(
+                    client, url, headers, params, method
+                )
+                response.raise_for_status()
+                return response.json(), dict(response.headers)
```

Comment:
```suggestion
                if self.refresh and self._has_token_expired(response.status_code):
                    await self.get_latest_token()
                    bitbucket_headers = await self._get_bitbucket_headers()
                    response = await self.execute_request(
                        client=client,
                        url=url,
                        headers=bitbucket_headers,
                        params=params,
                        method=method,
                    )
                response.raise_for_status()
                return response.json(), dict(response.headers)
```

This is a refresh protocol that is no-op for bring your own token, but is used in saas

See other comments regarding methods `_get_bitbuket_headers` and `_has_token_expired`

Reply:
✅ **Comment 7 Addressed**: Implemented the complete refresh protocol as suggested. The `_make_request()` method now includes token expiration detection using `_has_token_expired()`, calls `get_latest_token()` to refresh, updates headers via `_get_bitbucket_headers()`, and retries the request. This is a no-op when bringing your own token locally but enables SaaS functionality.

Reply:
✅ **Addressed**: Implemented the refresh protocol in `_make_request()` method (lines 116-125). The method now checks if `self.refresh` is enabled and if the token has expired (401 status), then calls `get_latest_token()` and `_get_bitbucket_headers()` to retry the request with fresh credentials. This is a no-op when bringing your own token locally but enables token refresh in SaaS environments.